### PR TITLE
LPS-75530 gradle plugin for osgi-bundle-builder to expose explodedJar task

### PR DESCRIPTION
**Pull request by david-truong**

### CODE REVIEW

**Comment by david-truong** [comment]
cc @gamerson

**Comment by david-truong** [comment]
@Ithildir it is not respecting compileInclude right now.. is that possible?

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	private void _addDependenciesOSGiBundleBuilder(Project project) {
+		GradleUtil.addDependency(
+			project, OSGI_BUNDLE_BUILDER_CONFIGURATION_NAME, "com.liferay",
+			"com.liferay.osgi.bundle.builder", "latest.release");
+	}
+
+	private ExplodedJarTask _addTaskExplodedJar(Project project) {
+		final SourceSet sourceSet = GradleUtil.getSourceSet(
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: Source sets are only available if the `java` plugin is applied. If someone applies this plugin without `java` (or applies this *before* `java`), the build will crash. I suggest applying the `java` plugin at the beginning of the `apply` method.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+			project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+		final SourceSetOutput sourceSetOutput = sourceSet.getOutput();
+
+		final ExplodedJarTask explodedJarTask = GradleUtil.addTask(
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: First create the task, then get the source set you need to configure it.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+			project, EXPLODED_JAR_TASK_NAME, ExplodedJarTask.class);
+
+		JavaCompile javaCompile = (JavaCompile)GradleUtil.getTask(
+			project, JavaPlugin.COMPILE_JAVA_TASK_NAME);
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: Use [this](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/SourceSet.html#getCompileJavaTaskName--) to get the task name.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+		Copy processResources = (Copy)GradleUtil.getTask(
+			project, JavaPlugin.PROCESS_RESOURCES_TASK_NAME);
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: Use [this](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/SourceSet.html#getProcessResourcesTaskName--) to get the task name.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+		explodedJarTask.dependsOn(javaCompile, processResources);
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: Actually, why don't you just depend on the `classes` task? Use [this](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/SourceSet.html#getClassesTaskName--) to get the task name.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+		explodedJarTask.setDescription(
+			"Builds your project as an exploded jar.");
+
+		explodedJarTask.setGroup(BasePlugin.BUILD_GROUP);
+
+		explodedJarTask.setBaseDir(
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: Please sort the setters, and leave empty lines only if the value is an action/callable/closure.

File: modules/sdk/gradle-plugins/src/main/java/com/liferay/gradle/plugins/LiferayOSGiPlugin.java [hunk 7]
```diff
@@ -199,6 +202,11 @@ public void execute(Project project) {
 			});
 	}
 
+	private void _addTaskExplodedJar(Project project) {
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins/src/main/java/com/liferay/gradle/plugins/LiferayOSGiPlugin.java]: Remove this, please 😋

File: modules/sdk/gradle-plugins-osgi-bundle-builder/build.gradle [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	into jar.destinationDir
+}
+
+dependencies {
+	compile group: "com.liferay", name: "com.liferay.gradle.util", version: "1.0.29"
+
+	compileOnly fileTree(builtBy: [rootProject.tasks.extractGradleApi25], dir: new File(rootProject.buildDir, "gradle-2.5"))
+}
+
+gradleTest {
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/build.gradle]: You added this code, but I don't see any Gradle tests. Can you please add a *smoke* test? Thanks!

File: modules/sdk/gradle-plugins-osgi-bundle-builder/build.gradle [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	dependsOn copyGradleTestDependencies
+	dependsOn jar
+
+	versions "2.5", "2.6", "2.7", "2.8", "2.9", "2.10", "2.11", "2.12", "2.13", "2.14.1", "3.0", "3.1", "3.2.1", "3.3"
+}
+
+//pluginBundle {
+//	tags = ["osgi"]
```
Comment (Ithildir) [inline modules/sdk/gradle-plugins-osgi-bundle-builder/build.gradle]: I'll take care of finding good tags for this plugin 😊

**Comment by Ithildir** [comment]
@david-truong, `compileInclude` is in `LiferayOSGiPlugin`, and it works by adding new Bnd instructions on the fly.

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/ExplodedJarTask.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+		super.exec();
+	}
+
+	@Optional
+	public File getBaseDir() {
+		return GradleUtil.toFile(getProject(), _baseDir);
+	}
+
+	@Input
```
[the remark this answers is not in the document]
Comment (david-truong) [reply modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/ExplodedJarTask.java]: Ahh that yes thanks!  I was wondering why things weren't deploying when I changed the bnd file

File: modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java (hunk 1 above)
Comment (david-truong) [reply modules/sdk/gradle-plugins-osgi-bundle-builder/src/main/java/com/liferay/gradle/plugins/osgi/bundle/builder/OSGiBundleBuilderPlugin.java]: ok!

**Comment by Ithildir** [comment]
@david-truong, I noticed several SF problems in the new Gradle plugin. Please run `gradlew formatSource` and fix them 😊

```
Added packageinfo for hello/world/constants: ./src/gradleTest/smoke/bnd.bnd (SourceCheck:BNDExportsCheck)
Incorrect Bundle-Name 'hello-world', see https://github.com/liferay/liferay-portal/blob/master/modules/util/source-formatter/documentation/bnd_bundle_information.markdown: ./src/gradleTest/smoke/bnd.bnd (SourceCheck:BNDBundleInformationCheck)
Incorrect Bundle-SymbolicName 'hello.world', see https://github.com/liferay/liferay-portal/blob/master/modules/util/source-formatter/documentation/bnd_bundle_information.markdown: ./src/gradleTest/smoke/bnd.bnd (SourceCheck:BNDBundleInformationCheck)
S:\lr\liferay-portal\modules\sdk\gradle-plugins-osgi-bundle-builder\src\gradleTest\smoke\src\main\resources\META-INF\resources\init.jsp (SourceCheck:JSPImportsCheck)
S:\lr\liferay-portal\modules\sdk\gradle-plugins-osgi-bundle-builder\src\gradleTest\smoke\src\main\resources\META-INF\resources\init.jsp (SourceCheck:JSPUnusedTermsCheck)
Missing copyright: ./src/gradleTest/smoke/src/main/resources/META-INF/resources/init.jsp (SourceCheck:CopyrightCheck)
S:\lr\liferay-portal\modules\sdk\gradle-plugins-osgi-bundle-builder\src\gradleTest\smoke\src\main\java\hello\world\portlet\HelloWorldPortlet.java (SourceCheck:JavaImportsCheck)
Missing copyright: ./src/gradleTest/smoke/src/main/java/hello/world/constants/HelloWorldPortletKeys.java (SourceCheck:CopyrightCheck)
Missing copyright: ./src/gradleTest/smoke/src/main/java/hello/world/portlet/HelloWorldPortlet.java (SourceCheck:CopyrightCheck)
S:\lr\liferay-portal\modules\sdk\gradle-plugins-osgi-bundle-builder\src\gradleTest\smoke\src\main\resources\META-INF\resources\view.jsp (SourceCheck:JSPTagAttributesCheck)
Missing copyright: ./src/gradleTest/smoke/src/main/resources/META-INF/resources/view.jsp (SourceCheck:CopyrightCheck)
Missing language key 'hello-world.caption': ./src/gradleTest/smoke/src/main/resources/META-INF/resources/view.jsp (SourceCheck:JSPLanguageKeysCheck)
public constant 'HelloWorld' of type 'String' must match pattern '^[A-Z0-9][_A-Z0-9]*$': ./src/gradleTest/smoke/src/main/java/hello/world/constants/HelloWorldPortletKeys.java 8 (Checkstyle:ConstantNameCheck)
```

**Comment by Ithildir** [comment]
@david-truong, also, you should (now or later) write the `README.markdown` for this new plugin.

**Pull request closed without merging**